什么是服务端请求伪造（SSRF）？ (#8886)
xitu/gold-miner: article/2021/server-side-request-forgery-vulnerability.md

> - 译文出自：[掘金翻译计划](https://github.com/xitu/gold-miner)
> - 本文永久链接：[https://github.com/xitu/gold-miner/blob/master/article/2021/server-side-request-forgery-vulnerability.md](https://github.com/xitu/gold-miner/blob/master/article/2021/server-side-request-forgery-vulnerability.md)
> - 译者：
> - 校对者：

# 什么是服务端请求伪造（SSRF）？

![](https://www.acunetix.com/wp-content/uploads/2019/02/ssrf_banner-910x273.png)

服务端请求伪造（SSRF）漏洞让攻击者从存在漏洞的应用程序的后端服务中发送恶意请求。攻击者通常使用 SSRF 攻击位于防火墙之后的内部系统，这些内部系统是不能从外部网络访问的。攻击者也可以利用 SSRF 访问通过服务器的环回接口（(127.0.0.1）可访问的服务。

当攻击者可以完全控制或部分控制应用程序发送的请求时，SSRF 漏洞就出现了。常见的例子是攻击者能控制第三方服务的 URL，并且应用程序将发起该 URL 对应的请求。

以下是一个 PHP 例子，它易遭受 SSRF 攻击。

```php
<?php

/**
* 检查是否设置了 GET 变量 'url'
* 例如 - http://localhost/?url=http://testphp.vulnweb.com/images/logo.gif
*/
if (isset($_GET['url'])){
$url = $_GET['url'];

/**
* 发起易遭受 SSRF 攻击的请求因为
* 在发起请求之前
* 没有对 $url 做验证
*/
$image = fopen($url, 'rb');

/**
* 发送正确的响应头
*/
header("Content-Type: image/png");

/**
* 响应图片的内容
*/
fpassthru($image);}
```

在上面的例子中，攻击者可以完全控制 **url** 参数。攻击者可以向因特网上的任意网站和被攻击服务器（**localhost**）上的资源发起 GET 请求。

在下面的例子中，攻击者向 Apache HTTP 服务发起一个请求。该 Apache HTTP 服务开启了 **mode_status** 模式（该模式默认是开启的）。

> ```

Reviewer: ```suggestion
```http request
```

GET /?url=http://localhost/server-status HTTP/1.1
Host: example.com
```

攻击者也可以使用 SSRF 向其他内部资源发起请求。这些内部资源可以被应用程序的服务器访问，但它们不是公开可访问的。例如，攻击者可以访问类似 AWS/[Amazon EC2](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) 和 [OpenStack](https://docs.openstack.org/admin-guide/compute-networking-nova.html) 的云服务实例的元数据。攻击者甚至可以利用 SSRF 发挥创意，[对内部 IP 进行端口扫描](https://www.acunetix.com/blog/articles/ssrf-vulnerability-used-to-scan-the-web-servers-network/)。

> ```

Reviewer: ```suggestion
```http request
```

> 攻击者也可以使用 SSRF 向其他内部资源发起请求。这些内部资源可以被应用程序的服务器访问，但它们不是公开可访问的。例如，攻击者可以访问类似 AWS/[Amazon EC2](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) 和 [OpenStack](https://docs.openstack.org/admin-guide/compute-networking-nova.html) 的云服务实例的元数据。攻击者甚至可以利用 SSRF 发挥创意，[对内部 IP 进行端口扫描](https://www.acunetix.com/blog/articles/ssrf-vulnerability-used-to-scan-the-web-servers-network/)。

Reviewer 2: ```suggestion
攻击者也可以使用 SSRF 向 web 服务器有权访问的其他内部资源发出请求，但它们不是公开可访问的。例如，攻击者可以访问类似 AWS/[Amazon EC2](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) 和 [OpenStack](https://docs.openstack.org/admin-guide/compute-networking-nova.html) 的云服务实例的元数据。攻击者甚至可以利用 SSRF 发挥创意，[对内部 IP 进行端口扫描](https://www.acunetix.com/blog/articles/ssrf-vulnerability-used-to-scan-the-web-servers-network/)。
```

GET /?url=http://169.254.169.254/latest/meta-data/ HTTP/1.1
Host: example.com
```

除了 **http://** 和 **https://** URL 协议之外，攻击者可以利用鲜为人知的或古老的 URL 协议访问本地系统或内部网络上的文件。

下面的例子使用了 **file:///** 协议。

> ```

Reviewer: ```suggestion
```http request
```

GET /?url=file:///etc/passwd HTTP/1.1
Host: example.com
```

部分应用可能允许攻击者使用更多奇怪的 URL 协议。比如，如果应用使用 cURL 发起请求，那么攻击者可以使用 **dict://** URL 协议向任意的主机和端口发起请求，并发送自定义数据。

> ```

Reviewer: ```suggestion
```http request
```

GET /?url=dict://localhost:11211/stat HTTP/1.1
Host: example.com
```

上面的请求使得应用程序连接 **localhost** 的 11211 端口，并发送 `stat` 字符串。端口 11211 是 [Memcached](https://memcached.org/) 的默认端口，该端口通常不会对外暴露。

## 检查服务端请求伪造

为了自动检测服务端请求伪造，你需要使用一个中间服务，因为检查这种漏洞需要一种外部可访问的并存在时延的方式。Acunetix 通过将 [AcuMonitor](https://www.acunetix.com/vulnerability-scanner/acumonitor-technology/) 作为中间服务解决该问题。

> 在扫描期间，Acunetix 向不同的 AcuMonitor URL 发起请求。如果 AcuMonitor 在这些不同的 URL 中收到一个请求，它将通知 Acunetix。然后 Acunetix 经发出 SSRF 警报。

Reviewer 2: ```suggestion
在扫描期间，Acunetix 向不同的 AcuMonitor URL 发起请求。如果 AcuMonitor 在这些不同的 URL 中收到唯一个请求，它将通知 Acunetix。然后 Acunetix 经发出 SSRF 警报。
```

Author: 原文的意思是说，收到任意一个请求就会发出 SSRF 警报。不是唯一个

下图是使用 AcuMonitor 的 Acunetix 扫描结果，它检查了服务端请求伪造。该警报包含 HTTP 请求的信息。它包括发起请求的服务器 IP 地址和请求中携带的 `User-Agent` 字符串（如果存在）。这些信息可以帮助开发者定位问题源头并修复问题。

![](https://www.acunetix.com/wp-content/uploads/2019/02/ssrf_screenshot-910x673.png)

## 消除服务端请求伪造

> 将简单的黑名单和正则表达式应用于用户输入是消除 SSRF 攻击的一种糟糕方法。通常，黑名单是一种几乎无效的安全控制手段，因为攻击者总是可以找到方法去绕过它们。在这个例子中，攻击者可以使用 HTTP 重定向、通配符 DNS 服务（比如 **xip.io**）或者甚至是 [可替代的 IP 编码](http://www.pc-help.org/obscure.htm)实现绕过黑名单。

Reviewer 2: ```suggestion
将简单的黑名单和正则表达式应用于用户输入是消除 SSRF 攻击的一种糟糕方法。通常，黑名单是一种几乎无效的安全控制手段。因为攻击者总是可以找到方法去绕过它们。在这个例子中，攻击者可以使用 HTTP 重定向、通配符 DNS 服务（比如 **xip.io**）或者甚至是 [可替代的 IP 编码](http://www.pc-help.org/obscure.htm)实现绕过黑名单。
```

### 白名单和 DNS 解析

> 最鲁棒的避免服务端请求伪造的方式是将你的应用需要访问的 DNS 名称和 IP 地址添加到白名单中。如果白名单机制不能满足你的需求，并且你必须依赖黑名单机制，那么对用户输入进行合理的校验就显得非常重要。比如，不允许向私有的 IP 地址发起请求（私用 IP 地址参考 [RFC 1918](https://tools.ietf.org/html/rfc1918)）。

Reviewer 2: ```suggestion
最稳健的避免服务端请求伪造的方式是将你的应用需要访问的 DNS 名称和 IP 地址添加到白名单中。如果白名单机制不能满足你的需求，并且你必须依赖黑名单机制，那么对用户输入进行合理的校验就显得非常重要。比如，不允许向私有的 IP 地址发起请求（私用 IP 地址参考 [RFC 1918](https://tools.ietf.org/html/rfc1918)）。
```
